### PR TITLE
Webhosting privatedatabase

### PR DESCRIPTION
**Pull request by wiseflat**
Hi,

This PR will add features to deploy privatesql databases as code.

### CODE REVIEW

**Review by pgaxatte — APPROVED** [review]
Awesome work, it's all good to me!

**Comment by scraly** [comment]
Hi,
thanks for your PR :).

Is it possible to fix your acceptance test in order to be able to execute all of them, without any order.
For example, in an acceptance test of a datasource you need to create the resource.
You can find a lot of examples of test acceptance of a datasource in the repository :).

File: website/docs/r/hosting_privatedatabase.html.markdown [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `description` - Custom description on your privatedatabase order.
+* `ovh_subsidiary` - (Required) Ovh Subsidiary
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase.html.markdown:56]:
```suggestion
* `ovh_subsidiary` - (Required) OVHcloud Subsidiary
```

File: website/docs/r/hosting_privatedatabase.html.markdown [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+* `payment_mean` - (Required) Ovh payment mode (One of "default-payment-mean", "fidelity", "ovh-account")
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase.html.markdown:57]:
```suggestion
* `payment_mean` - (Required) OVHcloud payment mode (One of "default-payment-mean", "fidelity", "ovh-account")
```

File: website/docs/r/hosting_privatedatabase.html.markdown [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `cpu` - Number of cpu on your private database
+* `datacenter` - Datacenter where this private database is located
+* `display_name` - Name displayed in customer panel for your private database
+* `hostname` - Private database hostname
+* `hostname_ftp` - Private database ftp hostname
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase.html.markdown:77]:
```suggestion
* `hostname_ftp` - Private database FTP hostname
```

File: website/docs/r/hosting_privatedatabase.html.markdown [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    * `label` - (Required) Identifier of the resource
+    * `value` - (Required) Path to the resource in API.OVH.COM
+
+Plan order valid values can be found on OVHcloud [APIv6](https://api.ovh.com/console/#/hosting/privateDatabase/availableOrderCapacities~GET)
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `cpu` - Number of cpu on your private database
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase.html.markdown:73]:
```suggestion
* `cpu` - Number of CPU on your private database
```

File: website/docs/r/hosting_privatedatabase_database.html.markdown [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+The following arguments are supported:
+
+* `service_name` - The internal name of your private database.
+* `database_name` - (Required) Name of your new database
+
+## Attributes Reference
+
+The id is set to the value of `service_name`/`database_name`.
+
+## Import
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase_database.html.markdown:33]: Please follow the same import block like other resources.
Example:

```

## Import

OVHcloud Managed Kubernetes Service clusters can be imported using the `serviceName` and the `id` of the cluster, separated by "/" E.g.,

```bash
$ terraform import ovh_cloud_project_kube.my_kube_cluster service_name/kube_id
```

File: website/docs/r/hosting_privatedatabase_user.html.markdown [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+* `service_name` - The internal name of your private database.
+* `password` - (Required) Password for the new user ( alphanumeric and 8 characters minimum )
+* `user_name` - (Required) User name used to connect on your databases
+
+## Attributes Reference
+
+The id is set to the value of `service_name`/`user_name`.
+
+## Import
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase_user.html.markdown:35]: Same remark, please follow the other import block example

File: website/docs/r/hosting_privatedatabase_user_grant.html.markdown [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  - admin
+  - none
+  - ro
+  - rw
+
+## Attributes Reference
+
+The id is set to the value of `service_name`/`user_name`/`database_name`/`grant` .
+
+## Import
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase_user_grant.html.markdown:41]: same remark for import block like others with an example

File: website/docs/r/hosting_privatedatabase_whitelist.html.markdown [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `service_name` - The internal name of your private database.
+* `ip` - (Required) The whitelisted IP in your instance.
+* `name` - (Required) Custom name for your Whitelisted IP.
+* `service` - (Required) Authorize this IP to access service port. Values can be `true` or `false`
+* `sftp` - (Required) Authorize this IP to access sftp port. Values can be `true` or `false`
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase_whitelist.html.markdown:33]:
```suggestion
* `sftp` - (Required) Authorize this IP to access SFTP port. Values can be `true` or `false`
```

File: website/docs/r/hosting_privatedatabase_whitelist.html.markdown [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+## Attributes Reference
+
+The id is set to the value of `service_name`/`ip`.
+
+## Import
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase_whitelist.html.markdown:39]: same remark :)

File: website/docs/d/hosting_privatedatabase.html.markdown [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+## Argument Reference
+
+* `service_name` - The service name
+
+## Attributes Reference
+
+`id` is set to database service_name. In addition, the following attributes are exported.
+
+* `cpu` - Number of cpu on your private database
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase.html.markdown:29]:
```suggestion
* `cpu` - Number of CPU on your private database
```

File: website/docs/d/hosting_privatedatabase.html.markdown [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+* `datacenter` - Datacenter where this private database is located
+* `display_name` - Name displayed in customer panel for your private database
+* `hostname` - Private database hostname
+* `hostname_ftp` - Private database ftp hostname
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase.html.markdown:33]:
```suggestion
* `hostname_ftp` - Private database FTP hostname
```

File: website/docs/d/hosting_privatedatabase.html.markdown [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+* `infrastructure` - Infrastructure where service was stored
+* `offer` - Type of the private database offer
+* `port` - Private database service port
+* `port_ftp` - Private database ftp port
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase.html.markdown:37]:
```suggestion
* `port_ftp` - Private database FTP port
```

File: website/docs/d/hosting_privatedatabase_database.html.markdown [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+## Argument Reference
+
+* `service_name` - The internal name of your private database
+* `database_name` - Database name
+
+## Attributes Reference
+
+`id` is set to `service_name`/`database_name`. In addition, the following attributes are exported.
+
+* `backup_time` - Time of the next backup ( every day )
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_database.html.markdown:31]:
```suggestion
* `backup_time` - Time of the next backup (every day)
```

File: website/docs/d/hosting_privatedatabase_whitelist.html.markdown [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+## Attributes Reference
+
+`id` is set to `service_name`/`ip`. In addition, the following attributes are exported.
+
+* `creation_date` - Creation date of the database
+* `last_update` - The last update date of this whitelist
+* `name` - Custom name for your Whitelisted IP
+* `service` - Authorize this IP to access service port
+* `sftp` - Authorize this IP to access sftp port
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_whitelist.html.markdown:35]:
```suggestion
* `sftp` - Authorize this IP to access SFTP port
```

File: ovh/data_hosting_privatedatabase.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+			"service_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			// Computed
+			"cpu": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Number of cpu on your private database",
```
Comment (scraly) [inline ovh/data_hosting_privatedatabase.go:23]:
```suggestion
				Description: "Number of CPU on your private database",
```

File: ovh/data_hosting_privatedatabase.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+			},
+			"hostname": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Private database hostname",
+			},
+			"hostname_ftp": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Private database ftp hostname",
```
Comment (scraly) [inline ovh/data_hosting_privatedatabase.go:43]:
```suggestion
				Description: "Private database FTP hostname",
```

File: ovh/data_hosting_privatedatabase.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+			},
+			"port": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Private database service port",
+			},
+			"port_ftp": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Private database ftp port",
```
Comment (scraly) [inline ovh/data_hosting_privatedatabase.go:63]:
```suggestion
				Description: "Private database FTP port",
```

File: ovh/data_hosting_privatedatabase_whitelist.go [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+			},
+			"service": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "Authorize this IP to access service port",
+			},
+			"sftp": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "Authorize this IP to access sftp port",
```
Comment (scraly) [inline ovh/data_hosting_privatedatabase_whitelist.go:49]:
```suggestion
				Description: "Authorize this IP to access SFTP port",
```

File: ovh/hosting_privatedatabase.go [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/ovh/go-ovh/ovh"
+)
+
+// WaitArchivedHostingPrivateDabaseTask wait for a task to become archived in the API (aka 404)
+func WaitArchivedHostingPrivateDabaseTask(client *ovh.Client, endpoint string, timeout time.Duration) error {
+	return resource.Retry(timeout, func() *resource.RetryError {
+		err := client.Get(endpoint, nil)
+		if errOvh, ok := err.(*ovh.APIError); ok && errOvh.Code != 404 {
+			log.Printf("[DEBUG][WaitArchivedHostingPrivateDabaseTask] weird error ... ")
```
Comment (scraly) [inline ovh/hosting_privatedatabase.go:17]: weird? :-D

File: ovh/hosting_privatedatabase.go [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+			return resource.NonRetryableError(err)
+		}
+		if err == nil {
+			log.Printf("[DEBUG][WaitArchivedHostingPrivateDabaseTask] WaitForArchived ... ")
+			return resource.RetryableError(errors.New("not archived yet"))
+		}
+		log.Printf("[DEBUG][WaitArchivedHostingPrivateDabaseTask] archived ! ")
```
Comment (scraly) [inline ovh/hosting_privatedatabase.go:24]:
```suggestion
		log.Printf("[DEBUG][WaitArchivedHostingPrivateDabaseTask] archived!")
```

File: ovh/hosting_privatedatabase.go (hunk 19 above)
Comment (scraly) [inline ovh/hosting_privatedatabase.go:17]:
```suggestion
			log.Printf("[DEBUG][WaitArchivedHostingPrivateDabaseTask] weird error ...")
```

File: ovh/import_hosting_privatedatabase_user_grant_test.go [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+resource "ovh_hosting_privatedatabase_user_grant" "grant" {
+    service_name  = "%s"
+    user_name     = "%s"
+    database_name = "%s"
+    grant         = "%s"
+}
+`
+
+func TestAccHostingPrivateDatabaseUserGrant_importBasic(t *testing.T) {
+	resourceName := "ovh_hosting_privatedatabase_user_grant.grant"
```
Comment (scraly) [inline ovh/import_hosting_privatedatabase_user_grant_test.go:22]: You need to add new environment variable in the documentation too:
https://registry.terraform.io/providers/ovh/ovh/latest/docs#testing-and-development

File: website/docs/d/hosting_privatedatabase_database.html.markdown [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+# ovh_hosting_privatedatabase_database (Data Source)
+
+Use this data source to retrieve information about an hosting privatedatabase.
+
+## Example Usage
+
+```hcl
+data "ovh_hosting_privatedatabase_database" "description" {
+ service_name  = "XXXXXX"
+ database_name  = "XXXXXX"
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_database.html.markdown:18]:
```suggestion
  database_name  = "XXXXXX"
```

File: website/docs/d/hosting_privatedatabase_user.html.markdown [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+# ovh_hosting_privatedatabase_user (Data Source)
+
+Use this data source to retrieve information about an hosting privatedatabase user.
+
+## Example Usage
+
+```hcl
+data "ovh_hosting_privatedatabase_user" "description" {
+ service_name  = "XXXXXX"
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_user.html.markdown:17]:
```suggestion
  service_name  = "XXXXXX"
```

File: website/docs/d/hosting_privatedatabase_user.html.markdown [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+ user_name  = "XXXXXX"
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_user.html.markdown:18]:
```suggestion
  user_name  = "XXXXXX"
```

File: website/docs/d/hosting_privatedatabase_user.html.markdown [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+---
+
+# ovh_hosting_privatedatabase_user (Data Source)
+
+Use this data source to retrieve information about an hosting privatedatabase user.
+
+## Example Usage
+
+```hcl
+data "ovh_hosting_privatedatabase_user" "description" {
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_user.html.markdown:16]:
```suggestion
data "ovh_hosting_privatedatabase_user" "user" {
```

File: website/docs/d/hosting_privatedatabase_user_grant.html.markdown [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+---
+
+# ovh_hosting_privatedatabase_user_grant (Data Source)
+
+Use this data source to retrieve information about an hosting privatedatabase user grant.
+
+## Example Usage
+
+```hcl
+data "ovh_hosting_privatedatabase_user_grant" "description" {
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_user_grant.html.markdown:16]:
```suggestion
data "ovh_hosting_privatedatabase_user_grant" "user_grant" {
```

File: website/docs/d/hosting_privatedatabase_user_grant.html.markdown [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+ service_name  = "XXXXXX"
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_user_grant.html.markdown:17]:
```suggestion
  service_name  = "XXXXXX"
```

File: website/docs/d/hosting_privatedatabase_user_grant.html.markdown [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+ database_name = "XXXXXX"
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_user_grant.html.markdown:18]:
```suggestion
  database_name = "XXXXXX"
```

File: website/docs/d/hosting_privatedatabase_user_grant.html.markdown [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+ user_name     = "XXXXXX"
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_user_grant.html.markdown:19]:
```suggestion
  user_name     = "XXXXXX"
```

File: website/docs/d/hosting_privatedatabase_whitelist.html.markdown [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+---
+
+# ovh_hosting_privatedatabase_whitelist (Data Source)
+
+Use this data source to retrieve information about an hosting privatedatabase whitelist.
+
+## Example Usage
+
+```hcl
+data "ovh_hosting_privatedatabase_whitelist" "description" {
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_whitelist.html.markdown:16]:
```suggestion
data "ovh_hosting_privatedatabase_whitelist" "whitelist" {
```

File: website/docs/d/hosting_privatedatabase_whitelist.html.markdown [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+ service_name  = "XXXXXX"
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_whitelist.html.markdown:17]:
```suggestion
  service_name  = "XXXXXX"
```

File: website/docs/d/hosting_privatedatabase_whitelist.html.markdown [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+ ip            = "XXXXXX"
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase_whitelist.html.markdown:18]:
```suggestion
  ip            = "XXXXXX"
```

File: website/docs/r/hosting_privatedatabase_user_grant.html.markdown [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+---
+
+# ovh_hosting_privatedatabase_user_grant
+
+Add grant on a database in your private cloud database instance.
+
+## Example Usage
+
+```hcl
+resource "ovh_hosting_privatedatabase_user_grant" "user" {
```
Comment (scraly) [inline website/docs/r/hosting_privatedatabase_user_grant.html.markdown:16]:
```suggestion
resource "ovh_hosting_privatedatabase_user_grant" "user_grant" {
```

File: website/docs/d/hosting_privatedatabase.html.markdown [hunk 34]
```diff
@@ -0,0 +1,45 @@
...
+
+```hcl
+data "ovh_hosting_privatedatabase" "database" {
+ service_name  = "XXXXXX"
+}
+```
+
+## Argument Reference
+
+* `service_name` - The service name
```
Comment (scraly) [inline website/docs/d/hosting_privatedatabase.html.markdown:23]:
```suggestion
* `service_name` - The internal name of your private database
```

**Review by scraly — CHANGES_REQUESTED** [review]

File: ovh/data_hosting_privatedatabase.go [hunk 35]
```diff
@@ -0,0 +1,157 @@
...
+			"version_number": {
+				Type:        schema.TypeFloat,
+				Computed:    true,
+				Description: "Private database version number",
+			},
+		},
+	}
+}
+
+func dataSourceHostingPrivateDatabaseRead(d *schema.ResourceData, meta interface{}) error {
```
Comment (scraly) [inline ovh/data_hosting_privatedatabase.go:114]: Small question:

What about setting the endpoint and dynamically the fields, like other datasources?
Example:

```
func dataSourcexxxxRead(d *schema.ResourceData, meta interface{}) error {
	config := meta.(*Config)
	serviceName := d.Get("service_name").(string)

	endpoint := fmt.Sprintf("/xx/xx/%s/kube/%s/xx", serviceName)
	res := &xxxxResponse{}

	log.Printf("[DEBUG] Will read xx from xx %s and xx: %s", xx, serviceName)
	err := config.OVHClient.Get(endpoint, res)
	if err != nil {
		return fmt.Errorf("calling get %s %w", endpoint, err)
	}
	for k, v := range res.ToMap() {
		if k != "id" {
			d.Set(k, v)
		}
	}
	d.SetId(xxId + "-" + res.ClientID + "-" + res.IssuerUrl)

	log.Printf("[DEBUG] Read xx %+v", res)
	return nil
} 
```

**Review by scraly — CHANGES_REQUESTED** [review]

**Review by scraly — APPROVED** [review]

**Comment by scraly** [comment]
Thanks! :)

**Pull request merged**